### PR TITLE
Added eslint module and configuration

### PR DESCRIPTION
**Pull request by ssarmis**
I hope I made the configurations good enough for #18 .
Also, I fixed some indentations in the build.js (I tested eslint on it).

### CODE REVIEW

**Comment by marco-c** [comment]
Thanks, looks good to me!

Could you make it run on Travis too?

**Comment by ssarmis** [comment]
@marco-c Should I also add a --fix when executing the check ?
Well, it wouldn't help, so I guess I shouldn't.

**Comment by ssarmis** [comment]
@marco-c Also, I get a lot of "no-undef" in your code, how should I deal with these ?
For example:
image

**Comment by marco-c** [comment]
> @marco-c Also, I get a lot of "no-undef" in your code, how should I deal with these ?

The problem is that those scripts are always loaded together with coverage.js.
The functions from coverage.js could be specified as globals in the configuration file (https://eslint.org/docs/user-guide/configuring#specifying-globals).
The environment for build.js should be `node` and the environment for the other scripts should be `webextensions`.

Perhaps it would be best to have two eslint configuration files, one for build.js with environment `node` and one for the other js scripts with environment `webextensions` and the globals definitions.
For simplicity, for now you could just have a single configuration file with both environments and globals definitions and use it for both build.js and all other js scripts.

**Comment by ssarmis** [comment]
Didn't find an easier way to make the functions/variables global, so I wrote the ones that were screaming.

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+module.exports = {
+    "globals": {
+        "document": true,
+        "window": true,
+        "MutationObserver": true,
+        "requestIdleCallback": true,
```
Comment (marco-c) [inline .eslintrc.js]: Aren't these four + `fetch` included in the `webextensions` env?

Comment (ssarmis) [reply .eslintrc.js]: No, I still get no-undef for them.

File: package.json [hunk 2]
```diff
@@ -3,10 +3,12 @@
   "version": "0.5.1",
   "description": "Extensions that augments several Mozilla websites (Bugzilla, DXR, Searchfox, MozReview) with code coverage information.",
   "dependencies": {
+    "eslint-plugin-mozilla": "^0.9.0",
     "web-ext": "^2.4.0"
   },
   "devDependencies": {},
   "scripts": {
+    "lint": "./node_modules/.bin/eslint **/*.js",
```
Comment (marco-c) [inline package.json]: You might be able to do just `eslint **/*.js` instead.

Comment (ssarmis) [reply package.json]: From what I experienced, I can't call ```eslint``` alone if its not globally installed.

Comment (ssarmis) [reply package.json]: Just tested, it works without the path, I tried when it wasn't in the package.json, my bad, I will remove the path.

File: .eslintrc.js (hunk 1 above)
Comment (marco-c) [reply .eslintrc.js]: Let's add the `browser` env too then, they should be defined in that one.

**Comment by ssarmis** [comment]
@marco-c Besides the path from the package.json file, is there any other change to be made ?

File: package.json [hunk 3]
```diff
@@ -3,10 +3,12 @@
   "version": "0.5.1",
   "description": "Extensions that augments several Mozilla websites (Bugzilla, DXR, Searchfox, MozReview) with code coverage information.",
   "dependencies": {
+    "eslint-plugin-mozilla": "^0.9.0",
     "web-ext": "^2.4.0"
   },
```
Comment (marco-c) [inline package.json]: `eslint` should be defined as a peerDependency: https://eslint.org/docs/developer-guide/working-with-plugins#peer-dependency. `eslint-plugin-mozilla` should probably be in devDependencies.
We should also move `web-ext` to `devDependencies`, but in a separate PR.

File: .eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
+module.exports = {
```
Comment (marco-c) [inline .eslintrc.js:1]: There should be an entry specifying that we want to use the `mozilla` plugin: `"plugins": ["mozilla"]`

File: .eslintrc.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    "extends": "eslint:recommended",
+    "parserOptions": {
+        "ecmaVersion": 8,
+        "ecmaFeatures": {
+            "jsx": true
+        },
+        "sourceType": "module"
+    },
+    "rules": {
```
Comment (marco-c) [inline .eslintrc.js:35]: Some of these rules might already be defined by `eslint-plugin-mozilla`, we should make sure we don't overwrite them.

Comment (marco-c) [reply .eslintrc.js:35]: To use those rules, I think we should add `extends: 'plugin:mozilla/recommended'` in the .eslintrc.

Comment (marco-c) [reply .eslintrc.js:35]: Here are the rules defined by default by the mozilla plugin: https://dxr.mozilla.org/mozilla-central/source/tools/lint/eslint/eslint-plugin-mozilla/lib/configs/recommended.js

File: .eslintrc.js (hunk 1 above)
Comment (marco-c) [reply .eslintrc.js]: Actually, the `browser` env might be implied automatically if we use the mozilla plugin's rules.

**Comment by ssarmis** [comment]
I will take a look tomorrow, its pretty late here.

**Comment by marco-c** [comment]
> I will take a look tomorrow, its pretty late here.

Sure, take your time ;)

**Comment by ssarmis** [comment]
I didn't remove rules like ```no-unused-vars``` or ```quotes``` because they were errors in the recommended and I let them be warnings.

File: .eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+module.exports = {
+    "globals": {
+        "waitIdle": true,
+        "wait": true,
+        "fetch": true,
```
Comment (marco-c) [inline .eslintrc.js:5]: Is `fetch` not already defined by default by the "browser" env from "mozilla/recommends"?

**Comment by marco-c** [comment]
> I didn't remove rules like no-unused-vars or quotes because they were errors in the recommended and I let them be warnings.

OK. We should remove the overwrites in follow-up PRs.

**Review by marco-c — APPROVED** [review]
Thanks!

**Comment by ssarmis** [comment]
I will remove ```fetch```, when I searched, I searched in the recommended file

**Comment by marco-c** [comment]
> I will remove fetch, when I searched, I searched in the recommended file

Oops, I merged the PR already. You can submit another PR to remove `fetch` (if it can indeed be removed).

**Comment by ssarmis** [comment]
I will make another pr then for the ```fetch```, eslint doesn't mind me removing it.

**Comment by marco-c** [comment]
Could you also remove all the overwritten rules, except the indent one (which is not enabled in the mozilla/recommended)?

**Pull request merged**